r/tools/nixlibs-allowlist.txt

```diff
@@ -2,3 +2,4 @@ ubuntu
 centos
 redhat
 rhel
+darwin
```

Comment:
Do we know if CRAN checks for downloads on macos? I know they do for some system types (hence the existence of this allowlist). But so far this allowlist was only platforms we knew weren't used by CRAN

Reply:
We can be sure that they don't as we previously used homebrew which downloads a copy of homebrew in addition to the binary bottles of the dependencies and cmake. So the number of downloads is reduced by this PR not increased ^^ (afaik we never added the local formula to the release packages as we do in the nightly packages so a source build wasn't possible)

Reply:
Just because they allowed downloading via (auto|home)brew doesn't mean they allow downloading binaries elsewhere. IIRC the autobrew stuff worked because CRAN agrees to it

> Only as a last resort and with the agreement of the CRAN team should a package download pre-compiled software. (See also [Using Rust](https://cran.r-project.org/web/packages/using_rust.html).)

from https://cran.r-project.org/web/packages/policies.html

It's possible that CRAN will treat `https://apache.jfrog.io/...` favorably, but looking at the recent experience of RUST (and listed in [Using Rust](https://cran.r-project.org/web/packages/using_rust.html, `github.com` is *not* considered stable enough (point being, it's not quite an all-or-nothing with downloading, there are some things that are allowed and some things that are not.

Reply:
> IIRC the autobrew stuff worked because CRAN agrees to it

I'm not certain that that's the case, but regardless, if the plan is to download binaries on CRAN, I agree with Jon's concerns, and we should think through some scenarios/options:

1. Just submit the release and don't say anything. Fine until it breaks, then you're flagged for violating the policy. You respond "ok, sorry, we'll stop" and fall back to building everything from source on CRAN. You'd need to make sure that all features like S3 were enabled in the mac bundled build on CRAN, otherwise people will find that things no longer work when they install binaries from CRAN. Maybe the long build times on CRAN machines will get them to concede to accept downloading binaries. Maybe the builds will fail because of weird system configuration they have that we aren't aware of. Maybe they'll archive the package for taking up too much of CRAN's resources on top of policy violation--generally being a "bad citizen" (I think I got that thrown at me once).

2. Request CRAN approval to download binaries for macOS. Explain that we have moved to a new build system that is fully under the ASF and Apache Arrow's governance, all releases are signed, voted on by the PMC, and checksummed, and that the ASF guarantees their hosting in perpetuity. And that we test for compatibility with system libraries before selecting and using a C++ binary. Hopefully CRAN finds the ASF sufficiently reliable and says yes. If they say yes, great. They may also ask "what about Linux and Windows"; we may be able to just do the same for both now too. They may say no, in which case we're all in build-from-source mode as above, and if that calls into question the Windows build, we have bigger problems because we don't have a bundled windows build today (which is why I might suggest only asking about downloading for macOS, and only after they approve then ask about other platforms).

Of the options, I personally lean to (2). It definitely has risks but the catastrophic risk is lower, I suspect. Plus, they do have in the policy that it is possible to get approval to download binaries, and I can't see how they could deny our request unless that's a hollow statement--ASF backing, open governance, thorough process, and (minus Windows for now) and completely possible, albeit slow, fallback to building from source.

Would love to hear others' thoughts.

Reply:
FWIW, I agree with Neal's lean towards (2) for similar reasons: the catastrophic outcomes are least likely with that option.

Reply:
> I understand the concerns and would agree with @nealrichardson that contacting CRAN and getting official approval is the best (if risky) way forward. With the bundled builds it should only be a matter of setting the feature flags to enable s3 and gcs for macos, everything else is covered by the default settings iirc.
> 

Correct; do you want to do that here or in a followup? This would also require some documentation changes to explain the different behavior between macOS and linux.

Or would we only want to make this change if CRAN rejects the binary downloading?

> With regards to discussions with CRAN: another positive to this new system in addition to the f/reliability aspects is the improved archive-ability: currently without autobrew available the src package can not be build on mac as we do not ship the package with the local formulas we use in the nightly builds to build from apache/arrow@main. With these changes that will be possible.
> 

Technically this is solvable with the current build system--we do for our nightly builds. 

> For linux an argument pro pre-compiled binary would be build/check times as arrow currently takes ~20x the checktime that dplyr takes and that is already on the longer side. So that would save cran a lot of resources and we could collaborate with them in someway to make sure one of the checks builds from source if they want that (that could just be setting `LIBARROW_BINARY=false`).
> 

I guess this is up to @thisisnic as the maintainer, but if it were me, I'd mention the possibility of doing Linux binary downloading but not necessarily couple it to the macOS change. There are what, 3, 4 macOS environments we have to build on? So many more bespoke machines to deal with for Linux, and while I think we've done all of the relevant checking to download a binary that will work, I wouldn't rule out that BDR has an evil machine somewhere that would break it, and I wouldn't want to have to fix that under time pressure. 

FWIW I don't think we'd need to tell them to set an env var to test the full source build. We already don't use the prebuilt binaries with `libc++` (https://github.com/apache/arrow/pull/37684/files/6a040aa4c55702999945b8e87c4bf572b2f7ec81#diff-935746c34b16289a07b0d9bf7642dbd268b18059b6187f7cdec7c464be47a3deR249) so that job would test the full source build. I'd stress this in the communication with CRAN: we don't blindly download binaries everywhere, only on systems that are compatible with the libraries we've built; for the rest, we build from source.

> For windows we are currently already hosting the binaries but are still using rwinlib primarily with our binaries as a fallback, we will need to switch that around for this release as well. I do want to extend the windows part of the build system to match linux and macos so that we can do source builds within the regular R package mechanisms but that won't be ready for the next release but could be mentioned if CRAN asks about windows.

+1

Reply:
> Correct; do you want to do that here or in a followup? 

I think a follow up, I agree with @paleolimbot, getting this merged first will allow some better testing and the PR is pretty big already.

Ah I wasn't aware that there is a job using libc++ , well that's perfect +1 on what you said :D

Reply:
> I guess this is up to @thisisnic as the maintainer, but if it were me, I'd mention the possibility of doing Linux binary downloading but not necessarily couple it to the macOS change.

Yeah, we will need to be careful about how we frame this, but if we can make things simpler/faster without complicating the discussion we need to have, let's give it a go.

Reply:
(I will leave darwin in the allowlist for now, we can adapt after _the talk_ with CRAN